Lib/test/test_zipimport.py

```
@@ -551,7 +551,12 @@ def replace(self, old, new):
             z.writestr(name, data)
             z.close()
             zi = zipimport.zipimporter(TEMP_ZIP)
-            self.assertEqual(data, zi.get_data(FunnyStr(name)))
+            try:
+                data2 = zi.get_data(FunnyStr(name))
+            except AttributeError:
```

What's the rationale for this change?  I can only guess that you're trying to ignore the error when `.get_data` is missing, but I don't know under what conditions that can happen with your branch.  Also, is it possible that something in the `FunnyStr()` call could raise the `AttributeError`?

If I'm guessing correctly, I'd like to see a comment, and to narrow the `except` clause.  Something like:

```
funny = FunnyStr(name)
try:
    # .get_data() could be missing because <fill in the reason>
    data2 = zi.get_data(funny)
except AttributeError:
    pass
```

---

`.get_data` is not missing. See [bpo-31291](https://bugs.python.org/issue31291). In the C implementation, passing the str subclass with the bad `replace()` method returning a non-string caused a crash (on Windows only). The simplest solution was to use the C equivalent of `str.replace(pathname, alt_path_sep, path_sep)` instead of `pathname.replace(alt_path_sep, path_sep)`. This always returns str. But for the Python implementation this issue doesn't exist. Passing a string with bad `relace()` doesn't cause a crash. It causes raising an AttributeError. I decided that the clearness of the Python code is more important than literal reproducing the C behavior in this case, and this behavior is good as well. Since I'm a co-author of this fix and this test, I think I can do this.

I'm not sure that this test still is needed.

---

It does seem like the test isn't needed.  I appreciate the high fidelity translation from C to Python, but I think in the cases where the C code had to protect against crashes, or did other things that aren't necessarily in Python, it makes sense to remove them in the Python version.  Within reason, it makes sense to take this opportunity to clean things up, IMHO.

---

Elsewhere @serhiy-storchaka said he wants to defer clean up to a second pass.  Let's put this issue in that category.